combined_iterator.go

```diff
@@ -0,0 +1,92 @@
+// Copyright © 2024 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/conduitio-labs/conduit-connector-mysql/common"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+)
+
+type combinedIterator struct {
+	snapshotIterator common.Iterator
+	cdcIterator      common.Iterator
+
+	currentIterator common.Iterator
+}
+
+type combinedIteratorConfig struct {
+	snapshotConfig snapshotIteratorConfig
+	cdcConfig      cdcIteratorConfig
+}
+
+func newCombinedIterator(
+	ctx context.Context,
+	config combinedIteratorConfig,
+) (common.Iterator, error) {
+	iterator := &combinedIterator{}
+	var err error
+
+	iterator.cdcIterator, err = newCdcIterator(ctx, config.cdcConfig)
```

Comment:
How does the CDC iterator know from whence to start streaming data?

Reply:
Currently in this pr it just always starts before the snapshot connector, and we assume that the window between the cdc iterator start and the snapshot iterator start is small enough. But yeah, that's not correct, events happening there could mess up the pipeline.

I guess the right way would be to capture the cdc position after the snapshot read lock has been acquired, and modify the `ErrSnapshotIteratorDone` to include the mysql position so that the cdc iterator can start.

We could join prs #23 and #24, with this change they are much more related. Or implement the correct version of this in pr #24, where I do the consistent snapshot.